src/Microsoft.ML.Recommender/MatrixFactorizationTrainer.cs

```diff
@@ -0,0 +1,329 @@
+        /// </summary>
+        /// <param name="env">The private instance of <see cref="IHostEnvironment"/>.</param>
+        /// <param name="labelColumn">The name of the label column.</param>
+        /// <param name="xColumnName">The name of the column hosting the matrix's column IDs.</param>
+        /// <param name="yColumnName">The name of the column hosting the matrix's row IDs.</param>
+        /// <param name="advancedSettings">A delegate to apply all the advanced arguments to the algorithm.</param>
+        /// <param name="context">The <see cref="TrainerEstimatorContext"/> for additional input data to training.</param>
+        public MatrixFactorizationTrainer(IHostEnvironment env, string labelColumn, string xColumnName, string yColumnName,
+            TrainerEstimatorContext context = null, Action<Arguments> advancedSettings = null)
+            : base(env, LoadNameValue)
+        {
+            var args = new Arguments();
+            advancedSettings?.Invoke(args);
+
+            _lambda = args.Lambda;
+            _k = args.K;
+            _iter = args.NumIterations;
+            _eta = args.Eta;
+            _threads = args.NumThreads ?? Environment.ProcessorCount;
+            _quiet = args.Quiet;
+            _doNmf = args.NonNegative;
+
+            Info = new TrainerInfo(normalization: false, caching: false);
+            Context = context;
+
+            LabelColumn = new SchemaShape.Column(labelColumn, SchemaShape.Column.VectorKind.Scalar, NumberType.R4, false);
+            XColumn = new SchemaShape.Column(xColumnName, SchemaShape.Column.VectorKind.Scalar, NumberType.U4, true);
+            YColumn = new SchemaShape.Column(yColumnName, SchemaShape.Column.VectorKind.Scalar, NumberType.U4, true);
+        }
+
+        /// <summary>
+        /// Train a matrix factorization model based on training data, validation data, and so on in the given context.
+        /// </summary>
+        /// <param name="context">The information collection needed for training. <see cref="TrainContext"/> for details.</param>
+        public override MatrixFactorizationPredictor Train(TrainContext context)
+        {
+            Host.CheckValue(context, nameof(context));
+
+            using (var ch = Host.Start("Training"))
+            {
+                var pred = TrainCore(ch, context.TrainingSet, context.ValidationSet);
+                return pred;
+            }
+        }
+
+        private MatrixFactorizationPredictor TrainCore(IChannel ch, RoleMappedData data, RoleMappedData validData)
+        {
+            Host.AssertValue(ch);
+            ch.AssertValue(data);
+            ch.AssertValueOrNull(validData);
+
+            ColumnInfo xColInfo;
+            ColumnInfo yColInfo;
+            ColumnInfo validXColInfo = null;
+            ColumnInfo validYColInfo = null;
+
+            Host.CheckValue(data.Schema.Label, nameof(data), "Input data did not have a unique label");
```

Comment:
>Host [](start = 12, length = 4)

use ch instead of host across whole function body.

Reply:
I don't understand. @singlis suggeted using Host for all cases.

Reply:
you need to use closest IExceptionContext available for you as possible.
Look on this this way:
Host responsible for whole trainer.
Channel responsible for this concreate train method.
What you prefer to see in exception: it failed somewhere in trainer or it failed in train channel?

---
In reply to: [227605504](https://github.com/dotnet/machinelearning/pull/1263#discussion_r227605504) [](ancestors = 227605504)

Reply:
Got it. I have changed `Host` to `ch` in `private MatrixFactorizationPredictor TrainCore(IChannel ch, RoleMappedData data, RoleMappedData validData)
` and `Host` to `env` in `public MatrixFactorizationTrainer(IHostEnvironment env, Arguments args) : base(env, LoadNameValue)`.